subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/VersionSelectionReasons.java

```diff
@@ -17,83 +17,127 @@
+        // Use a set to de-duplicate same descriptions
+        private final Set<ComponentSelectionDescriptorInternal> descriptions;
 
-        private DefaultComponentSelectionReason(boolean forced, boolean conflictResolution, boolean selectedByRule, boolean expected, boolean compositeBuild, String description) {
-            this.forced = forced;
-            this.conflictResolution = conflictResolution;
-            this.selectedByRule = selectedByRule;
-            this.expected = expected;
-            this.compositeParticipant = compositeBuild;
-            assert description != null;
-            this.description = description;
+        private DefaultComponentSelectionReason(ComponentSelectionDescriptor description) {
+            descriptions = Sets.newLinkedHashSet();
+            descriptions.add((ComponentSelectionDescriptorInternal) description);
+        }
+
+        public DefaultComponentSelectionReason(List<ComponentSelectionDescriptor> descriptions) {
+            this.descriptions = Sets.newLinkedHashSet();
+            for (ComponentSelectionDescriptor description : descriptions) {
+                this.descriptions.add((ComponentSelectionDescriptorInternal) description);
+            }
         }
 
 
         public boolean isForced() {
-            return forced;
+            return hasCause(ComponentSelectionCause.FORCED);
+        }
+
+        private boolean hasCause(ComponentSelectionCause cause) {
+            for (ComponentSelectionDescriptor description : descriptions) {
+                if (description.getCause() == cause) {
+                    return true;
+                }
+            }
+            return false;
         }
 
         public boolean isConflictResolution() {
-            return conflictResolution;
+            return hasCause(ComponentSelectionCause.CONFLICT_RESOLUTION);
         }
 
         public boolean isSelectedByRule() {
-            return selectedByRule;
+            return hasCause(ComponentSelectionCause.SELECTED_BY_RULE);
         }
 
         public boolean isExpected() {
-            return expected;
+            ComponentSelectionCause cause = Iterables.getLast(descriptions).getCause();
+            return cause == ComponentSelectionCause.ROOT || cause == ComponentSelectionCause.REQUESTED;
         }
 
         public boolean isCompositeSubstitution() {
-            return compositeParticipant;
+            return hasCause(ComponentSelectionCause.COMPOSITE_BUILD);
         }
 
         public String getDescription() {
-            return description;
+            // for backwards compatibility, we use the last added description
+            return Iterables.getLast(descriptions).toString();
         }
 
         public String toString() {
-            return description;
+            return getDescription();
+        }
+
+        @Override
+        public ComponentSelectionReasonInternal addCause(ComponentSelectionCause cause, String description) {
```

Comment:
I wonder if we wouldn't be better off making this type immutable (together with using an `ImmutableSet`), and create a new one in this instance. I guess it depends how frequently we expect to be adding causes to existing `ComponentSelectionReasonInternal`.

Immutable types are nice :).